File: src/main/java/org/nosphere/apache/rat/RatWork.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.nosphere.apache.rat;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.UncheckedIOException;
import java.lang.reflect.Field;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;
import org.apache.commons.io.function.IOSupplier;
import org.apache.rat.Defaults;
import org.apache.rat.ReportConfiguration;
import org.apache.rat.Reporter;
import org.apache.rat.analysis.IHeaderMatcher;
import org.apache.rat.analysis.TikaProcessor;
import org.apache.rat.analysis.matchers.OrMatcher;
import org.apache.rat.analysis.matchers.SimpleTextMatcher;
import org.apache.rat.api.Document;
import org.apache.rat.api.RatException;
import org.apache.rat.commandline.StyleSheets;
import org.apache.rat.license.ILicense;
import org.apache.rat.license.ILicenseFamily;
import org.apache.rat.license.LicenseSetFactory.LicenseFilter;
import org.apache.rat.report.claim.ClaimStatistic;
import org.gradle.api.GradleException;
import org.gradle.internal.logging.ConsoleRenderer;
import org.gradle.workers.WorkAction;

public abstract class RatWork implements WorkAction<RatWorkSpec> {

    private static final List<String> SCRIPT_MEDIA_TYPES = Arrays.asList(
            "application/x-sh", "application/x-bat", "application/javascript", "application/rls-services+xml");

    @Override
    public void execute() {
        restoreScriptDocumentTypes();
        RatWorkSpec spec = getParameters();
        File reportDir = spec.getReportDirectory().getAsFile().get();
        reportDir.mkdirs();
        ReportConfiguration config = configure(spec);
        Reporter reporter = new Reporter(config);
        ClaimStatistic stats = runAudit(reporter);
        report(reporter, reportDir);
        verdict(spec, config, stats, reporter, reportDir);
    }

    private static void verdict(
            RatWorkSpec spec, ReportConfiguration config, ClaimStatistic stats, Reporter reporter, File reportDir) {
        int unapproved = stats.getCounter(ClaimStatistic.Counter.UNAPPROVED);
        if (config.getClaimValidator().isValid(ClaimStatistic.Counter.UNAPPROVED, unapproved)) {
            return;
        }
        if (spec.getVerbose().get()) {
            System.err.println(unapprovedFilesListing(reporter));
        }
        String message = "Apache Rat audit failure - " + unapproved + " unapproved license"
                + (unapproved > 1 ? "s" : "") + "\n"
                + "\tSee " + new ConsoleRenderer().asClickableFileUrl(new File(reportDir, "index.html"));
        if (spec.getFailOnError().get()) {
            throw new GradleException(message);
        }
        System.err.println(message);
    }

    private static ClaimStatistic runAudit(Reporter reporter) {
        try {
            return reporter.execute();
        } catch (RatException ex) {
            throw new GradleException(ex.getMessage(), ex);
        }
    }

    private static void report(Reporter reporter, File reportDir) {
        writeReport(reporter, StyleSheets.XML.getStyleSheet(), new File(reportDir, "rat-report.xml"));
        writeReport(reporter, StyleSheets.PLAIN.getStyleSheet(), new File(reportDir, "rat-report.txt"));
        writeHtmlReport(reporter, new File(reportDir, "index.html"));
    }

    private static void writeReport(Reporter reporter, IOSupplier<InputStream> stylesheet, File target) {
        try {
            reporter.output(stylesheet, () -> new FileOutputStream(target));
        } catch (RatException ex) {
            throw new GradleException(ex.getMessage(), ex);
        }
    }

    private static String unapprovedFilesListing(Reporter reporter) {
        ByteArrayOutputStream listing = new ByteArrayOutputStream();
        try {
            reporter.output(StyleSheets.UNAPPROVED_LICENSES.getStyleSheet(), () -> listing);
        } catch (RatException ex) {
            throw new GradleException(ex.getMessage(), ex);
        }
        return new String(listing.toByteArray(), StandardCharsets.UTF_8);
    }

    private static void writeHtmlReport(Reporter reporter, File target) {
        try (InputStream stylesheet = RatWork.class.getResourceAsStream("apache-rat-output-to-html.xsl");
                OutputStream output = new FileOutputStream(target)) {
            Transformer transformer = TransformerFactory.newInstance().newTransformer(new StreamSource(stylesheet));
            transformer.transform(new DOMSource(reporter.getDocument()), new StreamResult(output));
        } catch (IOException ex) {
            throw new UncheckedIOException(ex);
        } catch (TransformerException ex) {
            throw new GradleException(ex.getMessage(), ex);
        }
    }

    @SuppressWarnings("unchecked")
    private static void restoreScriptDocumentTypes() {
        try {
            Field documentTypeMap = TikaProcessor.class.getDeclaredField("DOCUMENT_TYPE_MAP");
            documentTypeMap.setAccessible(true);
            Map<String, Document.Type> typesByMediaType = (Map<String, Document.Type>) documentTypeMap.get(null);
            for (String mediaType : SCRIPT_MEDIA_TYPES) {
                typesByMediaType.put(mediaType, Document.Type.STANDARD);
            }
        } catch (ReflectiveOperationException | RuntimeException ex) {
            throw new GradleException("Unable to register script media types as text documents with Apache Rat", ex);
        }
    }

    private static ReportConfiguration configure(RatWorkSpec spec) {
        ReportConfiguration config = new ReportConfiguration();
        config.setFrom(defaults(spec));
        List<SubstringMatcher> matchers = spec.getSubstringMatchers().get();
        registerSubstringMatchers(config, matchers);
        approveOnly(config, spec.getApprovedLicenses().get(), vocabulary(matchers));
        config.addSource(new FilesReportable(
                spec.getBaseDir().getAsFile().get(),
                new ArrayList<>(spec.getReportedFiles().getFiles())));
        return config;
    }

    private static String resolveApprovedLicense(LicenseFamilies knownFamilies, String value) {
        try {
            return knownFamilies.resolve(value);
        } catch (IllegalArgumentException ex) {
            throw configurationError("approvedLicenses: " + ex.getMessage());
        }
    }

    private static GradleException configurationError(String detail) {
        return new GradleException(
                "Apache Rat configuration error: " + detail + " failOnError does not apply to configuration errors.");
    }

    private static LicenseFamilies vocabulary(List<SubstringMatcher> matchers) {
        ReportConfiguration defaults = new ReportConfiguration();
        defaults.setFrom(Defaults.builder().build());
        Set<LicenseFamily> families = new TreeSet<>();
        for (ILicenseFamily family : defaults.getLicenseFamilies(LicenseFilter.ALL)) {
            families.add(LicenseFamily.of(family.getFamilyCategory(), family.getFamilyName()));
        }
        for (SubstringMatcher matcher : matchers) {
            families.add(LicenseFamily.of(matcher.getLicenseFamilyCategory(), matcher.getLicenseFamilyName()));
        }
        return new LicenseFamilies(families);
    }

    private static void approveOnly(
            ReportConfiguration config, List<String> approvedLicenses, LicenseFamilies knownFamilies) {
        if (approvedLicenses.isEmpty()) {
            return;
        }
        Set<String> approvedCategories = new TreeSet<>();
        for (String approvedLicense : approvedLicenses) {
            String category = resolveApprovedLicense(knownFamilies, approvedLicense);
            config.addApprovedLicenseCategory(category);
            approvedCategories.add(category);
        }
        Set<String> categoriesToRemove = new TreeSet<>(config.getLicenseCategories(LicenseFilter.APPROVED));
        categoriesToRemove.removeAll(approvedCategories);
        config.removeApprovedLicenseCategories(categoriesToRemove);
    }

    private static void registerSubstringMatchers(ReportConfiguration config, List<SubstringMatcher> matchers) {
        for (int index = 0; index < matchers.size(); index++) {
            SubstringMatcher matcher = matchers.get(index);
            String category = matcher.getLicenseFamilyCategory();
            String name = matcher.getLicenseFamilyName();
            if (!hasFamilyWithCategory(config, category)) {
                config.addFamily(ILicenseFamily.builder()
                        .setLicenseFamilyCategory(category)
                        .setLicenseFamilyName(name)
                        .build());
            }
            config.addLicense(ILicense.builder()
                    .setFamily(category)
                    .setName(name)
                    .setId(category.trim() + "-" + (index + 1))
                    .setMatcher(headerMatcherFor(matcher.getSubstrings())));
        }
    }

    private static boolean hasFamilyWithCategory(ReportConfiguration config, String category) {
        String paddedCategory = ILicenseFamily.makeCategory(category);
        for (ILicenseFamily family : config.getLicenseFamilies(LicenseFilter.ALL)) {
            if (family.getFamilyCategory().equals(paddedCategory)) {
                return true;
            }
        }
        return false;
    }

    private static IHeaderMatcher headerMatcherFor(List<String> substrings) {
        if (substrings.size() == 1) {
            return new SimpleTextMatcher(substrings.get(0));
        }
        List<IHeaderMatcher> textMatchers = new ArrayList<>();
        for (String substring : substrings) {
            textMatchers.add(new SimpleTextMatcher(substring));
        }
        return new OrMatcher(textMatchers, null);
    }

    private static Defaults defaults(RatWorkSpec spec) {
        Defaults.Builder builder = Defaults.builder();
        if (!spec.getAddDefaultMatchers().get()) {
            builder.noDefault();
        }
        return builder.build();
    }
}

File: src/main/kotlin/org/nosphere/apache/rat/RatWork.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.nosphere.apache.rat

import org.apache.rat.Defaults
import org.apache.rat.Report
import org.apache.rat.ReportConfiguration
import org.apache.rat.analysis.IHeaderMatcher
import org.apache.rat.analysis.util.HeaderMatcherMultiplexer
import org.apache.rat.anttasks.SubstringLicenseMatcher
import org.apache.rat.api.RatException
import org.apache.rat.document.impl.FileDocument
import org.apache.rat.license.SimpleLicenseFamily
import org.apache.rat.report.IReportable
import org.apache.rat.report.RatReport
import org.apache.rat.report.claim.ClaimStatistic
import org.apache.rat.report.xml.XmlReportFactory
import org.apache.rat.report.xml.writer.impl.base.XmlWriter
import org.gradle.api.file.ConfigurableFileCollection
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.Property

import org.gradle.internal.logging.ConsoleRenderer
import org.gradle.workers.WorkAction
import org.gradle.workers.WorkParameters

import org.w3c.dom.Element
import org.w3c.dom.NodeList

import java.io.File
import java.io.FilenameFilter
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.TransformerFactory
import javax.xml.transform.stream.StreamResult
import javax.xml.transform.stream.StreamSource

internal
abstract class RatWorkSpec : WorkParameters {
    abstract val verbose: Property<Boolean>
    abstract val failOnError: Property<Boolean>
    abstract val addDefaultMatchers: Property<Boolean>
    abstract val substringMatchers: ListProperty<SubstringMatcher>
    abstract val approvedLicenses: ListProperty<String>
    abstract val baseDir: DirectoryProperty
    abstract val reportedFiles: ConfigurableFileCollection
    abstract val excludeFile: RegularFileProperty
    abstract val stylesheet: RegularFileProperty
    abstract val reportDirectory: DirectoryProperty
}

internal
abstract class RatWork : WorkAction<RatWorkSpec> {

    override fun execute() {

        val reportDir = parameters.reportDirectory.asFile.get()
        reportDir.mkdirs()

        val xmlReportFile = reportDir.resolve("rat-report.xml")
        val plainReportFile = reportDir.resolve("rat-report.txt")
        val htmlReportFile = reportDir.resolve("index.html")

        val config = createReportConfiguration(parameters)

        val stats = ClaimStatistic()

        generateXmlReport(config, stats, xmlReportFile)

        transformReport(xmlReportFile, htmlReportFile, plainReportFile)

        if (stats.numUnApproved > 0) {
            if (parameters.verbose.get()) {
                System.err.println(verboseFailureOutput(xmlReportFile))
            }
            val message = "Apache Rat audit failure - " +
                "${stats.numUnApproved} unapproved license${if (stats.numUnApproved > 1) "s" else ""}\n" +
                "\tSee ${ConsoleRenderer().asClickableFileUrl(htmlReportFile)}"
            if (parameters.failOnError.get()) throw RatException(message)
            else System.err.println(message)
        }
    }

    private
    fun createReportConfiguration(spec: RatWorkSpec): ReportConfiguration =
        ReportConfiguration().apply {

            val matchers = mutableListOf<IHeaderMatcher>()
            if (spec.addDefaultMatchers.get()) matchers.add(Defaults.createDefaultMatcher())
            spec.substringMatchers.get().forEach { substringMatcher ->
                matchers.add(SubstringLicenseMatcher().apply {
                    licenseFamilyCategory = substringMatcher.licenseFamilyCategory
                    licenseFamilyName = substringMatcher.licenseFamilyName
                    substringMatcher.substrings.forEach { substring ->
                        addConfiguredPattern(SubstringLicenseMatcher.Pattern().apply {
                            this.substring = substring
                        })
                    }
                })
            }
            headerMatcher = HeaderMatcherMultiplexer(matchers)

            if (spec.approvedLicenses.get().isEmpty()) {
                isApproveDefaultLicenses = true
            } else {
                isApproveDefaultLicenses = false
                setApprovedLicenseNames(spec.approvedLicenses.get().map { SimpleLicenseFamily(it) })
            }
        }

    private
    fun generateXmlReport(config: ReportConfiguration, stats: ClaimStatistic, xmlReportFile: File) {

        xmlReportFile.bufferedWriter().use { xmlFileWriter ->
            val writer = XmlWriter(xmlFileWriter)
            XmlReportFactory.createStandardReport(writer, stats, config).run {
                startReport()
                FilesReportable(parameters.reportedFiles.files.toList(), parameters.excludeFile.asFile.orNull).run(this)
                endReport()
            }
            writer.closeDocument()
        }
    }

    private
    fun transformReport(xmlReportFile: File, htmlReportFile: File, plainReportFile: File) =
        TransformerFactory.newInstance().let { factory ->

            factory.newTransformer(StreamSource(parameters.stylesheet.asFile.get())).transform(
                StreamSource(xmlReportFile),
                StreamResult(htmlReportFile)
            )

            factory.newTransformer(StreamSource(Defaults.getPlainStyleSheet())).transform(
                StreamSource(xmlReportFile),
                StreamResult(plainReportFile)
            )
        }

    private
    fun verboseFailureOutput(xmlReportFile: File): String =
        unapprovedFilesFrom(xmlReportFile).joinToString(
            separator = "\n - ",
            prefix = "Files with unapproved licenses:\n - ",
            postfix = "\n"
        )

    private
    fun unapprovedFilesFrom(xmlReportFile: File): List<String> =
        DocumentBuilderFactory.newInstance().newDocumentBuilder()
            .parse(xmlReportFile)
            .getElementsByTagName("resource")
            .toElementList()
            .filter { resource ->
                resource.childNodes.toElementList().any {
                    it.tagName == "license-approval" && it.getAttribute("name") == "false"
                }
            }
            .map { it.getAttribute("name") }
            .sorted()

    private
    fun NodeList.toElementList(): List<Element> =
        ArrayList<Element>(length).apply {
            for (idx in 0 until length) {
                add(item(idx) as Element)
            }
        }
}

private
class FilesReportable(
    private val files: List<File>,
    private val excludeFile: File?
) : IReportable {

    override fun run(report: RatReport) {
        for (file in files.filter(excludeFileFilter())) {
            report.report(FileDocument(file))
        }
    }

    private
    fun excludeFileFilter(): (File) -> Boolean {
        val lines = excludeFile?.takeIf { it.isFile }?.readLines()?.filter { it.isNotBlank() }
        if (lines.isNullOrEmpty()) return { true }
        val filter = createFilenameFilter(lines)
        return { file -> filter.accept(file.parentFile, file.name) }
    }

    private
    fun createFilenameFilter(lines: List<String>) =
        Report::class.java.getDeclaredMethod("parseExclusions", List::class.java).run {
            isAccessible = true
            invoke(null, lines)
        } as FilenameFilter
}
